### PR TITLE
Improve GPT2

### PR DESCRIPTION
**Pull request by Toudsour**
Main changes:

- Get `vocabulary` and `merge` from attributes instead of environment
- Remove LRU cache
- Fix a bug: get contrary result when judging all space input
- Add some test cases
- Reformat some code according to clang-tidy's suggestions 
- Adjust the doc

### CODE REVIEW

File: ocos/kernels/kernels.h [hunk 1]
```diff
@@ -11,11 +11,13 @@ typedef CPTR_OrtCustomOp (*FxGetSchemaInstance)();
...
 
 struct BaseKernel {
-  BaseKernel(OrtApi api) : api_(api), ort_(api_) {}
+  BaseKernel(OrtApi api) : api_(api), info_(nullptr), ort_(api_) {}
+  BaseKernel(OrtApi api, const OrtKernelInfo *info) : api_(api), info_(info), ort_(api_) {}
 
  protected:
   OrtApi api_;  // keep a copy of the struct, whose ref is used in the ort_
   Ort::CustomOpApi ort_;
+  const OrtKernelInfo* info_;
```
Comment (wenbingl) [inline ocos/kernels/kernels.h:20]: where was 'info_' used?

Comment (xadupre) [reply ocos/kernels/kernels.h:20]: I reused this implementation in PR #41, info_ is needed to retrieve parameters.

**Comment by wenbingl** [comment]
> It's still WIP.

ready now?

**Comment by Toudsour** [comment]
> > It's still WIP.
> 
> ready now?

File: tokenizer/gpt2tok.cc [hunk 2]
```diff
@@ -97,22 +96,17 @@ using json = nlohmann::json;
...
+      : unk_id_(-1) {
   }
 
   struct BpeNode {
     int id;
     int value;
   };
 
-  void Load(const char* p_vocab_file, const char* p_bpe_file, const char* unk_token, const char* special_tokens) {
-    std::ifstream json_stream(p_vocab_file);
```
Comment (wenbingl) [inline tokenizer/gpt2tok.cc:109]: Did you guys verify the performance of  the method Load, c++ stream is not very performant in some case.

**Review by wenbingl — APPROVED** [review]

**Pull request merged**